File: apps/frontend/lib/api/scheduling.ts
/**
 * Reminders, interviews, and agenda API (P3 §E/§F/§G, Requirements 10–12).
 *
 * Reminders + interviews are nested under an application; the agenda is a merged,
 * time-ordered view across all applications. Creates accept an optional
 * idempotency key (double-submit safety). All calls are user-scoped server-side.
 */
import { apiFetch, apiPost, apiPatch, apiDelete, API_BASE } from './client';

// -- types ------------------------------------------------------------------

export type ReminderStatus = 'pending' | 'snoozed' | 'firing' | 'fired' | 'cancelled';
export type InterviewKind = 'screen' | 'technical' | 'onsite' | 'behavioral' | 'final' | 'other';

export interface Reminder {
  id: string;
  application_id: string;
  due_at: string;
  tz: string;
  note: string | null;
  recurrence: string | null;
  status: ReminderStatus;
  created_at: string;
  updated_at: string;
}

export interface ReminderCreate {
  due_at?: string;
  preset?: string;
  tz?: string;
  note?: string | null;
  recurrence?: string | null;
}

export interface OverlapWarning {
  id: string;
  starts_at: string;
}

export interface Interview {
  id: string;
  application_id: string;
  starts_at: string;
  tz: string;
  duration_min: number;
  kind: InterviewKind;
  location: string | null;
  notes: string | null;
  lead_times: number[];
  status: 'scheduled' | 'cancelled';
  created_at: string;
  updated_at: string;
  overlaps: OverlapWarning[];
}

export interface InterviewCreate {
  starts_at: string;
  tz?: string;
  duration_min?: number;
  kind?: InterviewKind;
  location?: string | null;
  notes?: string | null;
  lead_times?: number[];
}

export interface AgendaItem {
  kind: 'reminder' | 'interview';
  id: string;
  application_id: string;
  when: string;
  tz: string;
  title: string;
  status: string;
}

export interface AgendaResponse {
  items: AgendaItem[];
  next_cursor: string | null;
}

// -- helpers ----------------------------------------------------------------

async function asJson<T>(res: Response, fallback: string): Promise<T> {
  if (!res.ok) {
    const data = (await res.json().catch(() => ({}))) as { detail?: unknown };
    const detail = typeof data.detail === 'string' ? data.detail : null;
    throw new Error(detail || `${fallback} (status ${res.status}).`);
  }
  return res.json() as Promise<T>;
}

function idemHeaders(key?: string): Record<string, string> {
  return key ? { 'Idempotency-Key': key } : {};
}

// -- reminders --------------------------------------------------------------

export async function listReminders(applicationId: string): Promise<Reminder[]> {
  const res = await apiFetch(`/applications/${applicationId}/reminders`, {
    credentials: 'include',
  });
  if (res.status === 404) return [];
  return asJson<Reminder[]>(res, 'Failed to load reminders');
}

export async function createReminder(
  applicationId: string,
  payload: ReminderCreate,
  idempotencyKey?: string
): Promise<Reminder> {
  const res = await apiPost(`/applications/${applicationId}/reminders`, payload, undefined, {
    headers: idemHeaders(idempotencyKey),
  });
  return asJson<Reminder>(res, 'Failed to create reminder');
}

export async function updateReminder(
  applicationId: string,
  reminderId: string,
  payload: Partial<ReminderCreate>
): Promise<Reminder> {
  const res = await apiPatch(`/applications/${applicationId}/reminders/${reminderId}`, payload);
  return asJson<Reminder>(res, 'Failed to update reminder');
}

export async function snoozeReminder(
  applicationId: string,
  reminderId: string,
  payload: { until?: string; preset?: string }
): Promise<Reminder> {
  const res = await apiPost(
    `/applications/${applicationId}/reminders/${reminderId}/snooze`,
    payload
  );
  return asJson<Reminder>(res, 'Failed to snooze reminder');
}

export async function cancelReminder(applicationId: string, reminderId: string): Promise<void> {
  await asJson(
    await apiDelete(`/applications/${applicationId}/reminders/${reminderId}`),
    'Failed to cancel reminder'
  );
}

// -- interviews -------------------------------------------------------------

export async function listInterviews(applicationId: string): Promise<Interview[]> {
  const res = await apiFetch(`/applications/${applicationId}/interviews`, {
    credentials: 'include',
  });
  if (res.status === 404) return [];
  return asJson<Interview[]>(res, 'Failed to load interviews');
}

export async function createInterview(
  applicationId: string,
  payload: InterviewCreate,
  idempotencyKey?: string
): Promise<Interview> {
  const res = await apiPost(`/applications/${applicationId}/interviews`, payload, undefined, {
    headers: idemHeaders(idempotencyKey),
  });
  return asJson<Interview>(res, 'Failed to schedule interview');
}

export async function updateInterview(
  applicationId: string,
  interviewId: string,
  payload: Partial<InterviewCreate>
): Promise<Interview> {
  const res = await apiPatch(`/applications/${applicationId}/interviews/${interviewId}`, payload);
  return asJson<Interview>(res, 'Failed to update interview');
}

export async function cancelInterview(applicationId: string, interviewId: string): Promise<void> {
  await asJson(
    await apiDelete(`/applications/${applicationId}/interviews/${interviewId}`),
    'Failed to cancel interview'
  );
}

/** Absolute URL to the interview's ICS download (opened in a new tab). */
export function interviewIcsUrl(interviewId: string): string {
  return `${API_BASE}/interviews/${interviewId}.ics`;
}

// -- agenda -----------------------------------------------------------------

export async function getAgenda(cursor?: string, limit = 20): Promise<AgendaResponse> {
  const params = new URLSearchParams();
  if (cursor) params.set('cursor', cursor);
  params.set('limit', String(limit));
  const res = await apiFetch(`/agenda?${params.toString()}`, { credentials: 'include' });
  if (res.status === 404) return { items: [], next_cursor: null };
  return asJson<AgendaResponse>(res, 'Failed to load agenda');
}

File: apps/backend/app/schemas/scheduling.py
"""Pydantic schemas for reminders, interviews, and the agenda (P3 §E/§F/§G)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

ReminderStatus = Literal["pending", "snoozed", "firing", "fired", "cancelled"]
InterviewKind = Literal["screen", "technical", "onsite", "behavioral", "final", "other"]


# -- Reminders --------------------------------------------------------------


class ReminderCreate(BaseModel):
    """Create a reminder via an explicit UTC ``due_at`` or a quick ``preset``."""

    due_at: str | None = None
    preset: str | None = None
    tz: str = "UTC"
    note: str | None = Field(default=None, max_length=1000)
    recurrence: str | None = None


class ReminderUpdate(BaseModel):
    due_at: str | None = None
    tz: str | None = None
    note: str | None = Field(default=None, max_length=1000)
    recurrence: str | None = None


class ReminderSnooze(BaseModel):
    until: str | None = None
    preset: str | None = None


class ReminderResponse(BaseModel):
    id: str
    application_id: str
    due_at: str
    tz: str
    note: str | None = None
    recurrence: str | None = None
    status: ReminderStatus
    created_at: str
    updated_at: str


# -- Interviews -------------------------------------------------------------


class InterviewCreate(BaseModel):
    starts_at: str
    tz: str = "UTC"
    duration_min: int = Field(default=60, ge=1, le=1440)
    kind: InterviewKind = "screen"
    location: str | None = Field(default=None, max_length=500)
    notes: str | None = Field(default=None, max_length=2000)
    lead_times: list[int] | None = None


class InterviewUpdate(BaseModel):
    starts_at: str | None = None
    tz: str | None = None
    duration_min: int | None = Field(default=None, ge=1, le=1440)
    kind: InterviewKind | None = None
    location: str | None = Field(default=None, max_length=500)
    notes: str | None = Field(default=None, max_length=2000)
    lead_times: list[int] | None = None


class OverlapWarning(BaseModel):
    id: str
    starts_at: str


class InterviewResponse(BaseModel):
    id: str
    application_id: str
    starts_at: str
    tz: str
    duration_min: int
    kind: InterviewKind
    location: str | None = None
    notes: str | None = None
    lead_times: list[int]
    status: Literal["scheduled", "cancelled"]
    created_at: str
    updated_at: str
    # Soft warning on create/reschedule (never blocks) - empty when none.
    overlaps: list[OverlapWarning] = Field(default_factory=list)


# -- Agenda -----------------------------------------------------------------


class AgendaItem(BaseModel):
    kind: Literal["reminder", "interview"]
    id: str
    application_id: str
    when: str  # UTC ISO (due_at / starts_at)
    tz: str
    title: str
    status: str


class AgendaResponse(BaseModel):
    items: list[AgendaItem]
    next_cursor: str | None = None
